sssd_test_framework/utils/smartcard.py

```diff
@@ -0,0 +1,203 @@
+    ]
+
+    def __init__(self, host: MultihostHost, fs: LinuxFileSystem, svc: SystemdServices) -> None:
+        """
+        :param host: Multihost host.
+        :type host: MultihostHost
+        :param fs: Filesystem utility instance.
+        :type fs: LinuxFileSystem
+        :param svc: Systemd svc utility instance.
+        :type svc: SystemdServices
+        """
+        super().__init__(host)
+
+        self.cli: CLIBuilder = host.cli
+        """CLI builder utility for command construction."""
+
+        self.fs: LinuxFileSystem = fs
+        """Filesystem utility used to handle file operations."""
+
+        self.svc: SystemdServices = svc
+        """Systemd utility to manage and interact with svc."""
+
+    def init(self, label: str = "sc_test", so_pin: str = "12345678", user_pin: str = "123456") -> None:
+        """
+        Initializes a SoftHSM token with the given label and PINs.
+
+        Cleans cache directories and prepares the token directory.
+
+        :param label: Token label, defaults to "sc_test"
+        :type label: str, optional
+        :param so_pin: Security Officer PIN, defaults to "12345678"
+        :type so_pin: str, optional
+        :param user_pin: User PIN, defaults to "123456"
+        :type user_pin: str, optional
+        """
+        for path in self.OPENSC_CACHE_PATHS:
+            self.fs.rm(path)
+
+        self.fs.rm(self.TOKEN_STORAGE_PATH)
+        self.fs.mkdir(self.TOKEN_STORAGE_PATH)
+
+        args: CLIBuilderArgs = {
+            "label": (self.cli.option.VALUE, label),
+            "free": (self.cli.option.SWITCH, True),
+            "so-pin": (self.cli.option.VALUE, so_pin),
+            "pin": (self.cli.option.VALUE, user_pin),
+        }
+        command_line = self.cli.command("softhsm2-util --init-token", args)
+        self.host.conn.run(command_line, env={"SOFTHSM2_CONF": self.SOFTHSM2_CONF_PATH})
+
+    def add_key(self, key_path: str, key_id: str = "01", pin: str = "123456") -> None:
+        """
+        Adds a private key to the smart card.
+
+        :param key_path: Path to the private key.
+        :type key_path: str
+        :param key_id: Key ID, defaults to "01"
+        :type key_id: str, optional
+        :param pin: User PIN, defaults to "123456"
+        :type pin: str, optional
+        """
+        args: CLIBuilderArgs = {
+            "module": (self.cli.option.VALUE, "/usr/lib64/pkcs11/libsofthsm2.so"),
+            "login": (self.cli.option.SWITCH, True),
+            "pin": (self.cli.option.VALUE, pin),
+            "write-object": (self.cli.option.VALUE, key_path),
+            "type": (self.cli.option.VALUE, "privkey"),
+            "id": (self.cli.option.VALUE, key_id),
+        }
+        command_line = self.cli.command("pkcs11-tool", args)
+        self.host.conn.run(command_line, env={"SOFTHSM2_CONF": self.SOFTHSM2_CONF_PATH})
+
+    def add_cert(self, cert_path: str, cert_id: str = "01", pin: str = "123456") -> None:
```

Comment:
we can add another parameter to this function, ```private: bool | None = False```

Then in add_key function, call add_cert with a simple conditional in certs. If I'm not mistaken the only difference is "type". So

```
def add_cert(self, cert_path: str, cert_id: str = "01", pin: str = "123456", private: bool | None = False) -> None:
    ... blah blah ...

    if private: 
      "type": (self.cli.option.Value, "key")
    else:
      "type": (self.cli.option.Value, "cert") 


def add_key(self, key_path: str, key_id: str = "01", pin: str = "123456"):
    add_cert(key, key_id, pin, private=True)
```